Modules/_pickle.c

```diff
@@ -1678,25 +1678,30 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
 {
     /* Optional file methods */
     if (_PyObject_LookupAttr(file, &_Py_ID(peek), &self->peek) < 0) {
-        return -1;
+        goto error;
     }
     if (_PyObject_LookupAttr(file, &_Py_ID(readinto), &self->readinto) < 0) {
-        return -1;
+        goto error;
+    }
+    if (_PyObject_LookupAttr(file, &_Py_ID(read), &self->read) < 0) {
+        PyErr_Clear();
```

Comment:
No, we should not replace an arbitrary exception with a TypeError.
```suggestion
        goto error;
```

`PyErr_Clear()` almost always looks like a red herring to me. Normally it should only be called to clear a specified type of exception (after `PyErr_ExceptionMatches()`). If it is called for unspecified type of exception, it is only because we have no way to report an error due to the API limitations (and `PyErr_WriteUnraisable()` is more preferable here), or it does not make much sense to report an error, because we already at abnormal situation handling route.

Reply:
Thanks for the explanation. I would prefer to chain exceptions explicitly in `error`, then.

Reply:
I see now that your code from f320be77ffb73e3b9e7fc98c37b8df3975d84b40 is the best solution. I've updated the PR to be less intrusive. Thanks for your guidance and patience.